### DISCUSSION TITLE
State machine syntax

### DISCUSSION THREAD

**Discussion by utaal** [verus-lang/verus#12] [Language design]
This is to discuss proposed state machine syntax, for "regular" state machines (TLA+/Iron*) and for concurrent state machines.

cc/ @tjhance @jonhnet @Chris-Hawblitzel @odedp

**Comment by utaal** [comment]
These code samples record, in part, a discussion with @Chris-Hawblitzel and @jonhnet.

```rust
struct CrashSafeLog {
   persistent: LogSpec,
   ephemeral: LogSpec, // two "copies" of the same state machine (state)
   log: Map<nat, LogSpec>, // two "copies" of the same state machine (state)
 }

 impl CrashSafeLog {
     pub fn init(&self) {
         self.persistent.init()
         self.ephemeral == self.persistent
     }

     pub fn ephemeral_move(old, new)
     {
       // -> &&& persistent == persistent'

       &&& old.persistent == new.persistent
       // &&& old.ephemeral.next(new.epehemeral)
       &&& (old.ephemeral, new.ephemeral).next()

       &&& (old.log[i], new.log[i]).next()
       &&& action.log[i].next()

       // action.log[i]: (LogSpec, LogSpec)

       // disallow 
       // &&&
       // |||
       // &&&
     }
     // more appealing than
     pub fn ephemeral_move(old, new)
     {
       and([
         old.persistent == new.persistent,
         old.ephemeral(new),
       ])
     }

     // imperative
     // pub fn ephemeral_move(old, new) {
     //   new == Self {
     //     old.ephemeral.next(),
     //     ...old
     //   }
     // }

     new == Self {
         // TODO left here with Jon
         ephemeral: x :| action(old.ephemeral, x).next()
     }

     #[and] pub fn ephemeral_move(&old, &new)
     // affects ephemeral
     {
       old.persistent == new.persistent,
       old.ephemeral(new),
     }

     init predicate()
     {
         && persistent.init()
         && ephemeral == persistent
     }
 }

 // can we support both imperative and predicate for twostate
```

**Reply by utaal** [reply]
My older notes, with syntax proposals, and a small example:
https://github.com/Chris-Hawblitzel/rust/wiki/Language-Design,-State-Machine-Mockup

A mypyvy inspired attempt: https://github.com/Chris-Hawblitzel/rust/wiki/Language-Design,-State-Machine-Mypyvy-Rust

And @jaybosamiya's take on refinement using Traits:
https://github.com/Chris-Hawblitzel/rust/wiki/State-Machine-Verification-via-Rust-Traits

**Comment by tjhance** [comment]
Here are my notes for concurrent state machine features:

https://docs.google.com/document/d/1Y6sigqVg1eWRcj8-j9tdJPG11pQHjardASNhIARO4-g/edit

**Comment by utaal** [comment]
## Can we use the same/similar syntax for concurrent state machines and regular TLA+/Iron* state machines

**Meeting 2022/01/07**

Relationship between concurrent state machines (CSM) and TLA+/Iron* state machines. The question is: can we throw away the meaning for CSM and get an atomic state machines?

* in CSM, the transition updates are "imperative" (assignments)
  ```
  #[transition]
  inc_A {
	require !has_a_incremented;
	update counter’ := counter + 1;
	update has_a_incremented’ := true;
  }
  ```
  in CSM, you may sometime want
  ```
  update counter' such that predicate(counter, counter')
  ```
  that is, non-imperative.
  Travis suggests one could then add parameters:
  ```
  inc_A(s') {
        require !has_a_incremented; 
        update counter' predicate(s.counter, s'.counter)
	update has_a_incremented’ := true;
  }
  ```
  Are these options good/sufficient for regular atomic state machines?

* in regular state machines having a way to talk about "things that fall out of the invariant" would still be useful; it is in CSMs for the "transition" that doesn't update anything but has knowledge of the necessary shard.
  ```
  #[static]
  finish {
    require has_a_incremented == true;
    require has_b_incremented == true;
  }
  ```

Follow-up: what do atomic state machines look like using the same style as Travis' proposal.

**Comment by tjhance** [comment]
oh this is a small thing, but I really like the idea of declaring invariants via an attribute like this:

```rust
#[invariant]
fn is_good(&self) -> bool { !self.bad }

#[invariant]
fn no_cookies(&self) -> bool { self.cookies == 0 }
```

rather than directly defining an `inv()` predicates. The predicates marked `#[invariant]` would be automatically put together as an `inv()` predicate; that makes it easier to have named invariants without needing to write a big conjunct list.

**Comment by utaal** [comment]
**2022/01/11**

We seem to agree that we would like a mini language to describe state machines and transitions: a syntactic sugar for some "pure"-verus encoding. The desugared verus can be a `struct` representing the machine's state, and function representing the transitions; we probably also want to auto-generate a `next` predicate (any of the transitions).

The desugaring may look like the following:
```rust
struct StateMachine {
  a: u64,
}

impl StateMachine {
  fn increment(&self, &post) -> bool {
    post.a == self.a + 1
  }
}
```

@jonhnet asks how compounded state machines in the sugared state machine syntax?

We currently plan to do proofs (such as refinement) with the desugared syntax. @tjhance suggests some machinery to help with writing proofs for each of the refined state machine's transitions (to avoid the mega switch block).

Some of the machinery for concurrency may require type resolution, which would likely make this a compiler step "before" rust->VIR. I think we should desugar to a rust ast (maybe via a proc macro), so that we can still rely on verus/rustc typechecking and rust->VIR.

@tjhance: "I didn't hear a compelling example of something external to a state machine that would require complex desugaring, besides avoiding the mega switch to match transition. Concretely we should start with the state machine stuff, with some external interface, and then build the sugar."

@Chris-Hawblitzel: "simple state machines should be doable with macros", @utaal: "that still allows for nice error messages". The open technical question is how to do the desugaring for concurrent state machines.

**Reply by utaal** [reply]
Here's a more extensive example of de-sugared syntax, based on some old examples:

```rust
// state machine
pub struct LogSpec {
  pub log: Seq<Element>
}

impl LogSpec {
  pub fn pred_init(&self) -> bool
  {
    self.log == seq![]
  }

  pub fn step_query(&self /* pre */, post: &Self, idx: nat, result: Element) -> bool
  {
    0 <= idx && idx < self.log.len() &&
    result == self.log[idx] &&
    post == self
  }

  pub fn step_append(&self, post: &Self, element: Element) -> bool
  {
    post.log == self.log.append(element)
  }

  pub fn step_stutter(&self, post: &Self) -> bool
  {
    post == self
  }

  pub fn next(&self, post: &Self) -> bool
  {
     // ???
  }
}


// state machine
pub struct CrashSafeLog {
   pub persistent: LogSpec,
   pub ephemeral: LogSpec, // two "copies" of the same state machine (state)
}

impl CrashSafeLog {
   pub fn pred_init(&self) {
       self.persistent.pred_init() &&
       self.ephemeral.pred_init() &&
       self.ephemeral == self.persistent
   }

   pub fn step_ephemeral_move(&self /* pre */, post: &Self) -> bool
   {
      self.persistent == post.persistent &&
      self.ephemeral.next(post.ephemeral)
   }

   pub fn step_sync(&self, post: &Self) -> bool
   {
      post.persistent == self.ephemeral &&
      post.ephemeral == self.ephemeral
   }

   pub fn step_crash(&self, post: &Self) -> bool
   {
     post.ephemeral == self.persistent &&
     post.persistent == self.persistent
   }
  
   pub fn next(&self, post: &Self) -> bool
   {
     // ???
   }
 }
```

**Reply by utaal** [reply]
@jonhnet Jay normal form (an ADT with one variant per transition) would likely be required to be able to match on the transition in the proof.

That may look like this:
```rust
enum LogSpecStep {
  Query(idx: nat, result: Element),
  Append(element: Element),
  Stutter,
}

fn next_step(&self, post: &Self, step: LogSpecStep) {
  match step {
    Query(idx: nat, result: Element) => self.step_query(post, idx, result),
    Append(element: Element) => self.step_append(post, element),
    Stutter => self.stutter(post),
  }
}

fn next(&self, post: &Self) -> bool
{
  exists(|step: LogSpecStep| self.next_step(post, step));
}
```

**Reply by utaal** [reply]
@jonhnet `next` is a TLA idiom, but we are going to have compositions that don't use it as the "glue";

We're generating things anyway, so we may not need the help of `next` (when performing a transition of a component state machine).

The inner state machine `Step` becomes a _field_ of the outer state machine's `step`.

**Comment by utaal** [comment]
**2022/01/12**

At the current stage, we seem to agree that we discussed things where we had enough knowledge to make decisions: which syntactic style to use (and to use the same for concurrent and atomic machines), where to draw the line between custom syntax (the state machine definitions) and regular verus syntax (client and proofs). @tjhance notes that to understand what we can do syntax-wise, we need to answer the technical question of how to do the translation.

For concurrent state machine desugaring @tjhance believes we may not need type information and we may be able to encode sufficient type checks with proc macros by generating "assertions" that would fail to type-check (the downside being we don't/may not get nice error messages). His action plan is to try this out.

**Comment by tjhance** [comment]
Going to try to pull together some concrete notes for the architecture design.

First, we have some intermediate representation, "SMAST". For normal atomic state machines this looks something like:

```
SMAST = (seq<Field>, seq<TransitionFuncs>, seq<Invariant>)

TransitionFunc = fn name(args...) { stmt }

stmt = seq<stmt>
    | If expr { statement } else { statement }
    | Let(local_var_name, expr)
    | Update(field_name, expr)
    | Require(expr)
    | Assert(expr)
    | ... more for concurrent state machines
```

Here, I'll write SMAST as templated by the type of 'expr', which will either be `rust expr` or `VIR expr`. Thus we can have either `SMAST<Rust>` or `SMAST<VIR>`. (I think we can literally write the code this way, with SMAST as a templated type.)

There are two primary classes of 'thing' to generate:

 * Rust-level types
   * Includes state type, step type, type signatures of `next`, `inv`, the transition relations
 * VIR-level definitions and specs
   * Definition of `next` and `inv`
   * Internal obligations for well-formedness (invariant is preserved and asserts are satisfied)

The Rust-level types absolutely need to be generated at the Rust level. Thus, there are two potential pipelines:

Option 1
```
            proc_macro               proc_macro
Rust code  -----------> SMAST<Rust> -----------> type signatures & specs
                                                                     |
                                                                     | 
                                                                     | rust_verify
                                                                     | 
                                                                     | 
                                                                     v
                                                                    VIR
```

vs

Option 2
```
            proc_macro               proc_macro
Rust code  -----------> SMAST<Rust> -----------> type signatures
                          |   
                          |   
                          | rust_verify
                          |   
                          |   
                       SMAST<VIR> -------> VIR
```

If we really _really_ need to have type signatures depend on the `SMAST<VIR>`, we might be able to accomplish it by exporting these type signatures at the _crate_ level, without letting them be used internally by the crate. (Kinda like proc macros.) Let's consider this as a last-resort for now.

In either case, we need to able to generate the function signatures from `SMAST<Rust>`; that is, while treating the rust exprs as opaque Rust AST (i.e., without typechecking them). I'm confident we can do this for the atomic state machines, and I'm pretty sure we can do it for the concurrent state machines if we put enough structure in to the `SMAST<_>`.

What about generating the VIR specs? Should we generate the specs as Rust code from `SMAST<Rust>`, or as VIR code from `SMAST<VIR>`?

 * I think for atomic state machines, we can generate the specs fairly straightforwardly. 
 * For concurrent state machines, it is a little more complicated. We want to check that Exprs do not refer to fields that they "don't have access to" and this is much easier after type-checking (or at least, after name resolution): i.e., at the VIR level.
   * Generally, I expect `SMAST<VIR>` to be marginally more convenient.
 * However, if we go with Option 1, it's a little easier for the user to see what specs are being generated. In option 2, their crate would need to type-check before they could see that information.
   * If this is really a problem, users could structure their state machines per-crate.
 * We want to leave the option open for much fancier techniques, like say, exporting to IVy or mypyvy. This almost certainly requires the `SMAST<VIR>`.

Now that I've had time to write this out, I'm starting to think we should go with Option 2 (despite any conclusions drawn yesterday).

**Reply by tjhance** [reply]
update: option 2 didn't work out. currently the state_machine architecture is entirely in the macro

**Comment by tjhance** [comment]
### More about spec generation

Let's actually talk about the spec generation now. We need to generate a relational definition for each transition. This should be straightforward in the atomic case, because we're only using fairly standard constructs. Also, our fancy concurrency operations will expand to be in terms of `update`, `require`, and so on, so those are easily handled.

We also have to generate the "safety conditions", i.e., that the invariant is preserved across the transition, and that the asserts follow from the invariant. Again, standard enough.

```
WP(assert b; X)  :=  b ==> WP(X)
WP(assume b; X)  :=  b && WP(X)
```

You get the idea. (@odedp probably knows more about this)

### For proving the proof-obligations

Our framework should give the user the ability to add lemmas to the state machine definition to support the proof obligations. I'm thinking something like this:

```
#[transition]
pub fn foo(args...) { ... }

#[preserves_invariant(foo)]
pub fn foo_proof(args...) { /* prove invariant is preserved here */ }
```

Our framework will check that all the necessary proof obligations have been added. This should take care of the most common scenario of casing by transition-type to prove an invariant. (Luckily Rust doesn't require file dependencies to be acyclic within a crate, so we can still organize the invariant proofs into another file without any headache.)

We can actually get pretty fancy with this. For example, let's allow the user to specify invariants by name:

```
#[invariant]
pub fn invariant_1(args...) -> bool { ... }

#[invariant]
pub fn invariant_2(args...) -> bool { ... }
```

Then we can case easily on which invariant we want to prove:

```
#[preserves_invariant(foo, invariant_1)]
pub fn foo_proof_1(args...) { /* prove invariant is preserved here */ }

#[preserves_invariant(foo, invariant_2)]
pub fn foo_proof_1(args...) { /* prove invariant is preserved here */ }
```

**Comment by tjhance** [comment]
### On system state machines

I have a pretty solid idea of what I want out of tokenized concurrent state machines to serve as sync. primitives & guard protocols.

And I have a pretty good idea of what IO state machines will look like for asynchronous specs, at least when they are isolated programs with no external devices.

I have not very much thought on how to handle 'system state machines' (which is also the place where 'composite state machines' come into play). But if we want to make Verus THE programming language for systems verification, I think we definitely want to codify this process somehow. The challenge, I think, is not just in having an elegant way to compose state machines, **but in being able to tie the interface between the composed state machines to the _program_ interface in a principle and general way.** Honestly, this feels fairly tricky to me, and I don't have a clear idea of what I want yet. But I think this should be the driving question of our design decisions here.

Luckily, we have several case studies from IronFleet & VeriBetrKV ... unfortunately, we only have a single case study in the concurrency world. At some point, we should get together and have a code tour for that case study, so we can all be on the same page about the challenges here.

**Reply by tjhance** [reply]
One thing that happens in the cache example is that we have these state objects to represent "IO requests". This exists on both the (concurrent) cache state machine and the (atomic) disk state machine. In the composed model, there is a transition that can transfer a request across the interface boundary.

The IO request is represented as a red block, here:

<img width="424" alt="Screen Shot 2022-01-13 at 1 08 15 PM" src="https://user-images.githubusercontent.com/592054/149385923-c5643bd8-a943-4931-9ad6-5f07319f0dce.png">

Until now, we've almost always composed state machines by essentially taking their product and then adding some new transitions. In the above example, the IO request object in the cache represents "the right to perform the request". The act of actually performing the request via an API call is represented by the the transition given by the orange arrow, which puts on the request into the disk's IO queue.

However, this composition would be a lot simpler if the state machines had "overlapped state":

<img width="417" alt="Screen Shot 2022-01-13 at 1 08 18 PM" src="https://user-images.githubusercontent.com/592054/149386128-867bfa27-ebd7-42f5-a3ea-a86ca5deb238.png">

It's a lot easier if you don't need an extra transition to transfer some data from one field to another. Or have complex invariants about data being in one field or the other.

In the past, we've built composed state machines by saying that both machines take a step at the same time, with the steps tied together via some label variable. What if instead, we composed them so that they could communicate via shared state?

I'm now imagining an architecture where the individual state machines making up a system have "terminals" and a system is basically a collection of state machines where the terminals pair up in some compatible way.

**Reply by tjhance** [reply]
The trick is to make these fancy state-machines-with-interfaces composable.

For example, take the cache again. We have a system model (the Cache + the Disk) and this refines some abstract cache specification.

But we don't have a way to _use_ the cache specification. If we wanted to use the cache as part of a larger subsystem, we'd need to go back to the cache state machine, augment it with some more state, resulting in a more complicated system state machine, and do the refinement proof from there.

Instead, we'd like to be able to use the abstract cache spec as one building block in verifying the large system, where some component of the abstract cache spec serves as the interface-terminal. So for example, verifying the cache would mean:

We take as our bottom bread:

 * some trusted async IO interface to the disk, where the user can define a concurrent state machine with special operations representing IO requests

and as our top bread:

* an interface to the cache, where the user can define a concurrent state machine with special operations that operate on the ephemeral state

That top bread can then be used as the bottom bread in a verified key-value store, giving us a double decker sandwich. Maybe someone else will verify the bottom bread (i.e., prove the async IO interface is correct in terms of some lower-level sys calls).

(btw I think it's worth noting that the cache interface is basically a "separation logic interface" to the disk where individual tokens essentially express ownership over a 4kb block on the disk)

**Comment by tjhance** [comment]
State machine task tracking

- [ ] General state machines
  - [x] proof obligations for inductiveness & safety conditions
  - [ ] general state machine definition (via traits)
  - [ ] state machine refinement
  - [ ] story for composition & system architecture
- [ ] Concurrent/shardable state machines
  - [ ] Sharding Strategies
    - [x] variable
    - [x] constant
    - [x] not-tokenized
    - [x] option
    - [x] map
    - [x] multiset
    - [x] storage & guards
    - [ ] storage_multiset
    - [ ] generic monoid
    - [ ] operating with collections
  - [ ] IO state machines
  - [ ] Formalism
    - [ ] ???
  - [ ] Transition language
    - [ ] binding variables in -= ops
    - [ ] let-patterns / match statements

**Comment by utaal** [comment]
Following the demo, I had another look at the `state-machines` branch. It's very cool. I think it looks fairly rust-y without compromising on usability.

There are a couple syntax choices that I'd like to discuss,

From one of the examples:

```rust
state_machine!(
    X {
        fields {
            pub number: int,
        }

        #[init]
        fn initialize(&self) {
            update(number, 0);
        }

        ...

        #[invariant]
        pub fn is_bounded(&self) -> bool {
            self.number <= self.maximum
        }

        ...
    }
);
```

The way fields and functions are nested is a bit surprising to me (given how Rust looks like). 
I'm wondering if we may want to have something that resembles rust a bit more.

```rust
state_machine!(
    X {
        pub number: int,
    }

    impl {
        #[init]
        fn initialize ...

        #[transition]
        fn add ...
    }

    inv {
        #[invariant], ...
    }
);
```

I haven't looked into how these are parsed in the macros, so it may be that this would complicate things. Also, how the client code (which refers to state machine transitions/invariants) looks like probably influences this. Can you post the expansion (of this, or another state machine) you showed yesterday in a gist?

**Reply by tjhance** [reply]
adding extra nesting isn't hard technically. also i think adding the extra blocks could possibly help break it up and make it easier to read

**Comment by tjhance** [comment]
here's a gist with 2 examples and some expanded code https://gist.github.com/tjhance/09449928c193b729e7440d2da6213989

**Comment by utaal** [comment]
Implemented by @tjhance, mainly in https://github.com/verus-lang/verus/tree/main/source/state_machines_macros
